### PR TITLE
Clear Logs and View Logs Functionality Added

### PR DESCRIPTION
**Pull request by namithj**
# Pull Request

## What changed?

Clear Logs and View Logs Functionality Added

## Why did it change?

Make debugging easier and provide an easier way for casual users to see whats going on

## Did you fix any specific issues?

#62 

## CERTIFICATION

By opening this pull request, I do agree to abide by the [CODE OF CONDUCT](https://github.com/aspirepress/.github/CODE_OF_CONDUCT.md) and be bound by the terms of the [Contribution Guidelines](https://github.com/aspirepress/.github/CONTRIBUTING.md) in effect on the date and time of my contribution as proven by the revision information in GitHub. I also agree that any previous contributions shall be deemed subject to the terms of the ersion in effect on the date and time of this pull request, or any future revisions for pull requests I may submit.

Further, I certify that this work is my own, is original, does not violate the intellectual property of any other person or entity, and I am not violating any license agreements or contracts I have with any person or entity. Finally, I agree that this code may be licensed under any license deemed appropriate by AspirePress, including but not limited to open source, closed source, proprietary or custom licenses, and that such license terms neither violate my rights or my copyright to this code.

### CODE REVIEW

File: assets/css/aspire-update.css
Comment (costdev) [inline assets/css/aspire-update.css:43]: It looks like `.editorconfig` isn't being read in your IDE. The indent style for CSS is a single tab.

File: includes/class-controller.php
Comment (costdev) [inline includes/class-controller.php:60]:
```suggestion
		if ( isset( $_POST['nonce'] ) && wp_verify_nonce( $_POST['nonce'], 'aspireupdate-ajax' ) ) {
```
This should use `&&` rather than `||`.

🔢 Applies elsewhere in the PR.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:52]:
```suggestion
	 * @return boolean true on success and false on failure.
```
We can just use `boolean` here for the type. Also a minor typo correction.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:77]:
```suggestion
	 * @param integer $limit Max no of lines to return. Defaults to a 1000 lines.
```
Minor: Extra space was there.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:83]: We could early return here to reduce indentation. Just a suggestion though.

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for this @namithj! I've left some change request and a couple of suggestions. 🙂

**Comment by namithj** [comment]
Thanks, will update. I got another issue in this PR which I haven't been able to figure out.  I am getting an extra line break between each log entries in the viewer and not in the log file.  Can you take a look.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:92]: The extra line in the "View Log" screen is caused by the `PHP_EOL` here.

With it:
image


Without it:
image

Comment (namithj) [reply includes/class-debug.php:92]: It should be like image 1.  But I am getting another empty line there on Windows.

Comment (namithj) [reply includes/class-debug.php:92]: Image 1 is how I want it.

Comment (namithj) [reply includes/class-debug.php:92]: Can you try activating responses and requests too and see if you start to get two empty lines between each log entry.

Comment (costdev) [reply includes/class-debug.php:92]: This is what I see in the file:
image

With the `PHP_EOL` still there, I get this in the viewer:
image

After removing the `PHP_EOL`, I get this in the viewer:
image

Comment (namithj) [reply includes/class-debug.php]: This is happening because of an extra EOL being inserted by print_r.  We can leave it as it is for now.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:79]: Looking at this again, this method is written with the view log use case in mind. That's fine, though I think we can improve on this a bit.

1. The return value is escaped and translated.
    - This prevents potential processing before output. Sanitize Early, Escape Late (as possible)
    - Instead, the escaping can be done before the JSON response is set.
2. The return value is a string whether it's successful or not.
    - This makes it more difficult for the receiving code to determine _how_ the value should be used or presented.
    - If this is preferred to be a _pretty_ low-level method, we could return `string|false` and let the receiving code determine the message on failure.
    - The method determines the return value for an empty log file. Maybe it should just return an empty string and the receiving code can determine how to communicate that?
    - If this is preferred to be a high-level method, we can return `new WP_Error( $err_code, $err_message );` so errors can be identified and context can be understood. The receiving code can then check `is_wp_error( $return_value )` and use `return_value->get_error_message()`, appropriately escaping before output.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:89]: Similar opportunity for an early return and reduced indentation here.


```php
if ( ! $wp_filesystem->exists( $file_path ) || ! $wp_filesystem->is_readable( $file_path ) ) {
    return //...
}

$file_content = //...
//...
```

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:102]: Looks like double indentation here

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:139]: Add a `@return void` to this one.

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the updates @namithj!

I've left some requested changes for indent fixing or missing return value, as well as some thoughts about possible improvements for the `::read()` method.

File: includes/class-debug.php
Comment (namithj) [reply includes/class-debug.php:79]: I guess we should go the WP error class way for future benefits.  Will update the PR

Comment (costdev) [reply includes/class-debug.php:79]: Cool! When I get a chance, I'll take a look at the rest of the codebase to see if there's opportunities with other return values and open an issue if I find any, though I don't have any in mind at the moment.

File: assets/playground/blueprint.json
Comment (costdev) [inline assets/playground/blueprint.json:20]: Hold off on merging this PR until #173 is merged. Then merge `main` into this branch.

_Sorry for the disruption and inconvenience_

Comment (asirota) [reply assets/playground/blueprint.json]: #173 is merged.

**Review by asirota — APPROVED** [review]

File: assets/playground/blueprint.json
Comment (costdev) [reply assets/playground/blueprint.json]: Great! The PR still needs work but at least there won't be extra headaches come merge time now that 173 is merged. 🙂

**Comment by namithj** [comment]
Lets get this merged

**Comment by asirota** [comment]
> Lets get this merged

Have you addressed @costdev suggestions?

**Comment by namithj** [comment]
Yes, everything resolved

File: includes/class-controller.php
Comment (costdev) [inline includes/class-controller.php:83]:
```suggestion
					'content' => $content,
```

**Review by costdev — CHANGES_REQUESTED** [review]

File: includes/class-controller.php
Comment (costdev) [inline includes/class-controller.php:79]:
```suggestion
				$content = $content->get_error_message();
       
```
Shouldn't we send a JSON error here instead?

Comment (namithj) [reply includes/class-controller.php:79]: We are showing those messages inside the viewer itself, if we throw json error it will go into an alert box.

Comment (namithj) [reply includes/class-controller.php:79]: On second thought I guess it make sense to throw those in the alert box. Let's reopen it

File: includes/class-controller.php
Comment (namithj) [reply includes/class-controller.php:83]: Thanks for spotting this.

**Comment by namithj** [comment]
@costdev PR Updated.  Added more error handling too.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:121]: These are already checked above so we can remove these checks and go straight to the `$wp_filesystem->put_contents()` call.

File: includes/class-debug.php
Comment (costdev) [inline includes/class-debug.php:154]: Looks like we may need to verify that the path is writable here, though I may be mistaken.

Comment (namithj) [reply includes/class-debug.php:154]: It will prevent us from writing to the file the first time ie file creation

Comment (namithj) [reply includes/class-debug.php:154]: We can check if the write operation is successful but we don't have any place to show the result

**Pull request merged**
